packer crash

#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue

packer crash, after removing some sections from a working hcl file.

#### Reproduction Steps

unfortunately, I lost the repro while trying to reduce the input file.

### Packer version

```
$ packer version
Packer v1.12.0
```

### Simplified Packer Template


### Operating system and Environment details

ubuntu 24.04

### Log Fragments and crash.log files

```
$ cat crash.log~
2025/02/10 10:26:28 [INFO] Packer version: 1.12.0 [go1.22.9 linux amd64]
2025/02/10 10:26:28 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2025/02/10 10:26:28 [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/hanwen/.packerconfig
2025/02/10 10:26:28 [WARN] Config file doesn't exist: /home/hanwen/.packerconfig
2025/02/10 10:26:28 [INFO] Setting cache directory: /home/hanwen/.cache/packer
2025/02/10 10:26:28 [TRACE] listing potential installations for <nil> that match "". plugingetter.ListInstallationsOptions{PluginDirectory:"/home/hanwen/.config/packer/plugins", BinaryInstallationOptions:plugingetter.BinaryInstallationOptions{APIVersionMajor:"5", APIVersionMinor:"0", OS:"linux", ARCH:"amd64", Ext:"", Checksummers:[]plugingetter.Checksummer{plugingetter.Checksummer{Type:"sha256", Hash:(*sha256.digest)(0xc0004e4080)}}, ReleasesOnly:false}}
2025/02/10 10:26:28 [INFO] found external [chroot ebs ebssurrogate ebsvolume instance] builders from amazon plugin
2025/02/10 10:26:28 [INFO] found external [import] post-processors from amazon plugin
2025/02/10 10:26:28 found external [ami parameterstore secretsmanager] datasource from amazon plugin
2025/02/10 10:26:28 plugin "/home/hanwen/.config/packer/plugins/github.com/hashicorp/amazon/packer-plugin-amazon_v1.3.4_x5.0_linux_amd64" does not support Protobuf, forcing use of Gob
2025/02/10 10:26:28 [INFO] found external [-packer-default-plugin-name-] builders from docker plugin
2025/02/10 10:26:28 [INFO] found external [import push save tag] post-processors from docker plugin
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_base_image_windows_2022_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for prerelease, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_debian12_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_ubuntu2404_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for version, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_customers, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_builder_project, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_windows_2022_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_builder_zone, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_builder_zone_arm, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_base_image_debian12_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_base_image_debian12_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for script_debug, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_ubuntu2004_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_macos14_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_ubuntu2004_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for gcp_base_image_debian11_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_debian11_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_macos14_arm64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_builder_region, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_debian12_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_macos_builder_region, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_xxxxx_debian_11_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_debian11_x86_64, so skipping
2025/02/10 10:26:28 [TRACE] validateValue: not active for aws_base_image_ubuntu2404_x86_64, so skipping
2025/02/10 10:26:28 waiting for all plugin processes to complete...
panic: interface conversion: interface is nil, not dag.Vertex

goroutine 1 [running]:
github.com/hashicorp/packer/internal/dag.stronglyConnected(0xc0008f88e0, 0xc000527300, {0x1771840, 0xc0006dd9a0})
	github.com/hashicorp/packer/internal/dag/tarjan.go:31 +0x4fb
github.com/hashicorp/packer/internal/dag.StronglyConnected(0xc000527300)
	github.com/hashicorp/packer/internal/dag/tarjan.go:19 +0x267
github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Cycles(0x5041a6?)
	github.com/hashicorp/packer/internal/dag/dag.go:62 +0x19
github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Validate(0xc000527300)
	github.com/hashicorp/packer/internal/dag/dag.go:34 +0x2a
github.com/hashicorp/packer/hcl2template.(*PackerConfig).buildPrereqsDAG(0xc00058c000)
	github.com/hashicorp/packer/hcl2template/parser.go:446 +0x732
github.com/hashicorp/packer/hcl2template.(*PackerConfig).evaluateBuildPrereqs(0xc00058c000, 0x0)
	github.com/hashicorp/packer/hcl2template/parser.go:459 +0x69
github.com/hashicorp/packer/hcl2template.(*PackerConfig).Initialize(0xc00058c000, {0x60?, 0x2f?})
	github.com/hashicorp/packer/hcl2template/parser.go:524 +0x32e
github.com/hashicorp/packer/command.(*BuildCommand).RunContext(0xc0005dede0, {0x1fea400, 0xc000540140}, 0xc000736000)
	github.com/hashicorp/packer/command/build.go:105 +0xed
github.com/hashicorp/packer/command.(*BuildCommand).Run(0xc0005dede0, {0xc00003e0c0, 0x3, 0x3})
	github.com/hashicorp/packer/command/build.go:45 +0xbb
github.com/mitchellh/cli.(*CLI).Run(0xc00058cf00)
	github.com/mitchellh/cli@v1.1.5/cli.go:262 +0x5b8
main.wrappedMain()
	github.com/hashicorp/packer/main.go:274 +0xeac
main.realMain()
	github.com/hashicorp/packer/main.go:51 +0xd5
main.main()

```

I am facing the same bug when adding a `local` depending on a `data` block:

This fails:

```hcl
data "http" "trusted_ca_certificates" {
  method = "GET"
  url    = "http://example.com/ca-bundle.crt"
}

locals {
  test = data.toto_trusted_ca_certificates.url
}
```

This works without a crash:

```hcl
data "http" "trusted_ca_certificates" {
  method = "GET"
  url    = "http://example.com/ca-bundle.crt"
}
```

```shell
packer inspect $(PACKER_VARIABLES) my-vmi
```

```shell
panic: interface conversion: interface is nil, not dag.Vertex

goroutine 1 [running]:
github.com/hashicorp/packer/internal/dag.stronglyConnected(0x1400091ee08, 0x140008146c0, {0x105acbb40, 0x140006795e0})
        github.com/hashicorp/packer/internal/dag/tarjan.go:31 +0x458
github.com/hashicorp/packer/internal/dag.StronglyConnected(0x140008146c0)
        github.com/hashicorp/packer/internal/dag/tarjan.go:19 +0x1f4
github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Cycles(0x1400091efe8?)
        github.com/hashicorp/packer/internal/dag/dag.go:62 +0x20
github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Validate(0x140008146c0)
        github.com/hashicorp/packer/internal/dag/dag.go:34 +0x28
github.com/hashicorp/packer/hcl2template.(*PackerConfig).buildPrereqsDAG(0x14000600140)
        github.com/hashicorp/packer/hcl2template/parser.go:446 +0x560
github.com/hashicorp/packer/hcl2template.(*PackerConfig).evaluateBuildPrereqs(0x14000600140, 0x1)
        github.com/hashicorp/packer/hcl2template/parser.go:459 +0x6c
github.com/hashicorp/packer/hcl2template.(*PackerConfig).Initialize(0x14000600140, {0x0?, 0x0?})
        github.com/hashicorp/packer/hcl2template/parser.go:524 +0x2c4
github.com/hashicorp/packer/command.(*ValidateCommand).RunContext(0x14000599380, {0x14000138020?, 0x3?}, 0x1400077e080)
        github.com/hashicorp/packer/command/validate.go:84 +0x124
github.com/hashicorp/packer/command.(*ValidateCommand).Run(0x14000599380, {0x14000138020, 0x3, 0x3})
        github.com/hashicorp/packer/command/validate.go:29 +0xac
github.com/mitchellh/cli.(*CLI).Run(0x14000523900)
        github.com/mitchellh/cli@v1.1.5/cli.go:262 +0x4a8
main.wrappedMain()
        github.com/hashicorp/packer/main.go:274 +0xc40
main.realMain()
        github.com/hashicorp/packer/main.go:51 +0xb8
main.main()
        github.com/hashicorp/packer/main.go:37 +0x1c



!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Packer crashed! This is always indicative of a bug within Packer.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Packer[1] so that we can fix this.

[1]: https://github.com/hashicorp/packer/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
make: *** [packer-lint] Error 2
```

> I am facing the same bug when adding a `local` depending on a `data` block:
> 
> This fails:
> 
> data "http" "trusted_ca_certificates" {
>   method = "GET"
>   url    = "http://example.com/ca-bundle.crt"
> }
> 
> locals {
>   test = data.trusted_ca_certificates.url
> }
> This works without a crash:
> 
> data "http" "trusted_ca_certificates" {
>   method = "GET"
>   url    = "http://example.com/ca-bundle.crt"
> }
> packer inspect $(PACKER_VARIABLES) my-vmi
> panic: interface conversion: interface is nil, not dag.Vertex
> 
> goroutine 1 [running]:
> github.com/hashicorp/packer/internal/dag.stronglyConnected(0x1400091ee08, 0x140008146c0, {0x105acbb40, 0x140006795e0})
>         github.com/hashicorp/packer/internal/dag/tarjan.go:31 +0x458
> github.com/hashicorp/packer/internal/dag.StronglyConnected(0x140008146c0)
>         github.com/hashicorp/packer/internal/dag/tarjan.go:19 +0x1f4
> github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Cycles(0x1400091efe8?)
>         github.com/hashicorp/packer/internal/dag/dag.go:62 +0x20
> github.com/hashicorp/packer/internal/dag.(*AcyclicGraph).Validate(0x140008146c0)
>         github.com/hashicorp/packer/internal/dag/dag.go:34 +0x28
> github.com/hashicorp/packer/hcl2template.(*PackerConfig).buildPrereqsDAG(0x14000600140)
>         github.com/hashicorp/packer/hcl2template/parser.go:446 +0x560
> github.com/hashicorp/packer/hcl2template.(*PackerConfig).evaluateBuildPrereqs(0x14000600140, 0x1)
>         github.com/hashicorp/packer/hcl2template/parser.go:459 +0x6c
> github.com/hashicorp/packer/hcl2template.(*PackerConfig).Initialize(0x14000600140, {0x0?, 0x0?})
>         github.com/hashicorp/packer/hcl2template/parser.go:524 +0x2c4
> github.com/hashicorp/packer/command.(*ValidateCommand).RunContext(0x14000599380, {0x14000138020?, 0x3?}, 0x1400077e080)
>         github.com/hashicorp/packer/command/validate.go:84 +0x124
> github.com/hashicorp/packer/command.(*ValidateCommand).Run(0x14000599380, {0x14000138020, 0x3, 0x3})
>         github.com/hashicorp/packer/command/validate.go:29 +0xac
> github.com/mitchellh/cli.(*CLI).Run(0x14000523900)
>         github.com/mitchellh/cli@v1.1.5/cli.go:262 +0x4a8
> main.wrappedMain()
>         github.com/hashicorp/packer/main.go:274 +0xc40
> main.realMain()
>         github.com/hashicorp/packer/main.go:51 +0xb8
> main.main()
>         github.com/hashicorp/packer/main.go:37 +0x1c
> 
> 
> 
> !!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
> 
> Packer crashed! This is always indicative of a bug within Packer.
> A crash log has been placed at "crash.log" relative to your current
> working directory. It would be immensely helpful if you could please
> report the crash with Packer[1] so that we can fix this.
> 
> [1]: https://github.com/hashicorp/packer/issues
> 
> !!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
> make: *** [packer-lint] Error 2

Actually my code is invalid, I am missing the `.http` in the reference, this is the valid code:

```hcl
data "http" "trusted_ca_certificates" {
  method = "GET"
  url    = "http://example.com/ca-bundle.crt"
}

locals {
  test = data.http.trusted_ca_certificates.url
}
```

--- a/hcl2template/parser.go
+++ b/hcl2template/parser.go
@@ -9,6 +9,7 @@ import (
 	"path/filepath"
 	"reflect"
 
+	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/go-version"
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/ext/dynblock"
@@ -398,6 +399,8 @@ func (cfg *PackerConfig) buildPrereqsDAG() (*dag.AcyclicGraph, error) {
 
 	verticesMap := map[string]dag.Vertex{}
 
+	var err error
+
 	// Do a first pass to create all the vertices
 	for ref := range cfg.Datasources {
 		// We keep a reference to the datasource separately from where it
@@ -429,27 +432,48 @@ func (cfg *PackerConfig) buildPrereqsDAG() (*dag.AcyclicGraph, error) {
 	for _, ds := range cfg.Datasources {
 		dsName := fmt.Sprintf("data.%s", ds.Name())
 
+		source := verticesMap[dsName]
+		if source == nil {
+			err = multierror.Append(err, fmt.Errorf("unable to find source vertex %q for dependency analysis, this is likely a Packer bug", dsName))
+			continue
+		}
+
 		for _, dep := range ds.Dependencies {
-			retGraph.Connect(
-				dag.BasicEdge(verticesMap[dsName],
-					verticesMap[dep.String()]))
+			target := verticesMap[dep.String()]
+			if target == nil {
+				err = multierror.Append(err, fmt.Errorf("unknown dependency %q for %q", dep.String(), dsName))
+				continue
+			}
+
+			retGraph.Connect(dag.BasicEdge(source, target))
 		}
 	}
 	for _, loc := range cfg.LocalBlocks {
 		locName := fmt.Sprintf("local.%s", loc.LocalName)
 
+		source := verticesMap[locName]
+		if source == nil {
+			err = multierror.Append(err, fmt.Errorf("unable to find source vertex %q for dependency analysis, this is likely a Packer bug", locName))
+			continue
+		}
+
 		for _, dep := range loc.dependencies {
-			retGraph.Connect(
-				dag.BasicEdge(verticesMap[locName],
-					verticesMap[dep.String()]))
+			target := verticesMap[dep.String()]
+
+			if target == nil {
+				err = multierror.Append(err, fmt.Errorf("unknown dependency %q for %q", dep.String(), locName))
+				continue
+			}
+
+			retGraph.Connect(dag.BasicEdge(source, target))
 		}
 	}
 
-	if err := retGraph.Validate(); err != nil {
-		return nil, err
+	if validateErr := retGraph.Validate(); validateErr != nil {
+		err = multierror.Append(err, validateErr)
 	}
 
-	return &retGraph, nil
+	return &retGraph, err
 }
 
 func (cfg *PackerConfig) evaluateBuildPrereqs(skipDatasources bool) hcl.Diagnostics {

--- a/hcl2template/parser.go
+++ b/hcl2template/parser.go
@@ -441,7 +441,7 @@ func (cfg *PackerConfig) buildPrereqsDAG() (*dag.AcyclicGraph, error) {
 		for _, dep := range ds.Dependencies {
 			target := verticesMap[dep.String()]
 			if target == nil {
-				err = multierror.Append(err, fmt.Errorf("unknown dependency %q for %q", dep.String(), dsName))
+				err = multierror.Append(err, fmt.Errorf("could not get dependency %q for %q, %q missing in template", dep.String(), dsName, dep.String()))
 				continue
 			}
 
@@ -461,7 +461,7 @@ func (cfg *PackerConfig) buildPrereqsDAG() (*dag.AcyclicGraph, error) {
 			target := verticesMap[dep.String()]
 
 			if target == nil {
-				err = multierror.Append(err, fmt.Errorf("unknown dependency %q for %q", dep.String(), locName))
+				err = multierror.Append(err, fmt.Errorf("could not get dependency %q for %q, %q missing in template", dep.String(), locName, dep.String()))
 				continue
 			}
 

--- a/packer_test/dag_tests/malformed_dep_test.go
+++ b/packer_test/dag_tests/malformed_dep_test.go
@@ -1,0 +1,49 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/packer/packer_test/common/check"
+)
+
+type malformedDepTestCase struct {
+	name          string
+	command       string
+	templatePath  string
+	useSequential bool
+}
+
+func genMalformedDepTestCases() []malformedDepTestCase {
+	retVals := []malformedDepTestCase{}
+
+	for _, cmd := range []string{"build", "validate"} {
+		for _, template := range []string{"./templates/malformed_data_dep.pkr.hcl", "./templates/malformed_local_dep.pkr.hcl"} {
+			for _, seq := range []bool{true, false} {
+				retVals = append(retVals, malformedDepTestCase{
+					name: fmt.Sprintf("Malformed dep packer %s --use-sequential-evaluation=%t %s",
+						cmd, seq, template),
+					command:       cmd,
+					templatePath:  template,
+					useSequential: seq,
+				})
+			}
+		}
+	}
+
+	return retVals
+}
+
+func (ts *PackerDAGTestSuite) TestMalformedDependency() {
+	pluginDir := ts.MakePluginDir()
+	defer pluginDir.Cleanup()
+
+	for _, tc := range genMalformedDepTestCases() {
+		ts.Run(tc.name, func() {
+			ts.PackerCommand().UsePluginDir(pluginDir).
+				SetArgs(tc.command,
+					fmt.Sprintf("--use-sequential-evaluation=%t", tc.useSequential),
+					tc.templatePath).
+				Assert(check.MustFail())
+		})
+	}
+}

--- a/packer_test/dag_tests/templates/malformed_data_dep.pkr.hcl
+++ b/packer_test/dag_tests/templates/malformed_data_dep.pkr.hcl
@@ -1,0 +1,8 @@
+data "http" "trusted_ca_certificates" {
+  method = "GET"
+  url    = "http://example.com/ca-bundle.crt"
+}
+
+locals {
+  test = data.trusted_ca_certificates.url
+}

--- a/packer_test/dag_tests/templates/malformed_local_dep.pkr.hcl
+++ b/packer_test/dag_tests/templates/malformed_local_dep.pkr.hcl
@@ -1,0 +1,4 @@
+data "http" "trusted_ca_certificates" {
+  method = "GET"
+  url    = local.no_dep
+}

